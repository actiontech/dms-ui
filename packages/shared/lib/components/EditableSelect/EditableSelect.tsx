import { useState, useMemo, forwardRef } from 'react';
import { Dropdown, Menu, Space, Popconfirm, Spin } from 'antd';
import { BasicButton } from '../BasicButton';
import { BasicInput } from '../BasicInput';
import { useTranslation } from 'react-i18next';
import {
  EditableSelectStyleWrapper,
  EditableSelectTriggerStyleWrapper
} from './style';
import { EditableSelectProps, EditableSelectOption } from './index.type';
import classNames from 'classnames';
import { EmptyBox } from '../EmptyBox';
import {
  PlusOutlined,
  MinusCircleFilled,
  EditFilled,
  UpOutlined,
  DownOutlined
} from '@actiontech/icons';
import type { MenuProps } from 'antd';
import { ReminderInformation } from '../ReminderInformation';

const EditableSelect: React.FC<EditableSelectProps> = ({
  value,
  onChange,
  onAdd,
  onUpdate,
  onDelete,
  options = [],
  addButtonText,
  placeholder,
  disabled = false,
  deletionConfirmTitle,
  loading = false,
  addable = true,
  updatable = true,
  deletable = true,
  errorMessage,
  searchable = true,
  searchPlaceholder,
  contentMaxHeight = 320
}) => {
  const { t } = useTranslation();
  const [newItemName, setNewItemName] = useState('');
  const [isAdding, setIsAdding] = useState(false);
  const [editingItem, setEditingItem] = useState<EditableSelectOption | null>(
    null
  );
  const [editName, setEditName] = useState('');
  const [dropdownOpen, setDropdownOpen] = useState(false);
  const [deletingItem, setDeletingItem] = useState<EditableSelectOption | null>(
    null
  );
  const [searchValue, setSearchValue] = useState('');

  const handleAdd = () => {
    onAdd?.(newItemName);
    setNewItemName('');
    setIsAdding(false);
  };

  const handleUpdate = () => {
    onUpdate?.({
      value: editingItem?.value ?? '',
      label: editName.trim()
    });
    setEditingItem(null);
    setEditName('');
  };

  const startEditing = (item: EditableSelectOption) => {
    setEditingItem(item);
    setEditName(item.label);
    setDeletingItem(null);
  };

  const cancelEditing = () => {
    setEditingItem(null);
    setEditName('');
  };

  const startAdding = () => {
    setIsAdding(true);
    setDeletingItem(null);
  };

  const resetState = () => {
    setNewItemName('');
    setIsAdding(false);
    setEditingItem(null);
    setEditName('');
    setDeletingItem(null);
    setSearchValue('');
  };

  const onOpenChange = (open: boolean) => {
    if (!open) {
      resetState();
    }
    setDropdownOpen(open);
  };

  const filteredOptions = useMemo(() => {
    if (!searchable || !searchValue.trim()) {
      return options;
    }

    const searchTerm = searchValue.toLowerCase().trim();
    return options.filter((option) =>
      option.label.toLowerCase().includes(searchTerm)
    );
  }, [searchable, searchValue, options]);

  const generateMenuItems = (): MenuProps['items'] => {
    const optionItems: MenuProps['items'] = filteredOptions.map((option) => ({
      key: option.value,
      className: classNames({
        'editable-select-menu-item-active': option?.value === value
      }),
      onClick: () => {
        onChange?.(option.value);
        setDropdownOpen(false);
        resetState();
      },
      label: (
        <div>
          <div className="menu-item-content">
            {editingItem?.value === option.value ? (
              <div className="edit-mode" onClick={(e) => e.stopPropagation()}>
                <BasicInput
                  value={editName}
                  autoFocus
                  onChange={(e) => setEditName(e.target.value)}
                />
                <div className="button-group">
                  <BasicButton size="small" onClick={cancelEditing}>
                    {t('common.cancel')}
                  </BasicButton>
                  <BasicButton
                    type="primary"
                    size="small"
                    onClick={handleUpdate}
                    disabled={!editName.trim()}
                  >
                    {t('common.save')}
                  </BasicButton>
                </div>
              </div>
            ) : (
              <>
                <span>{option.label}</span>
                <Space>
                  <EmptyBox if={updatable}>
                    <BasicButton
                      type="text"
                      size="small"
                      icon={<EditFilled />}
                      onClick={(e) => {
                        e.stopPropagation();
                        startEditing(option);
                      }}
                      className="edit-button"
                    />
                  </EmptyBox>
                  <EmptyBox if={deletable}>
                    <Popconfirm
                      title={
                        deletionConfirmTitle ?? t('common.deleteConfirmTitle')
                      }
                      onConfirm={(e) => {
                        e?.stopPropagation();
                        setDeletingItem(option);
                        onDelete?.(option);
                      }}
                      onCancel={(e) => {
                        e?.stopPropagation();
                      }}
                    >
                      <BasicButton
                        type="text"
                        size="small"
                        danger
                        icon={<MinusCircleFilled width={14} height={14} />}
                        onClick={(e) => {
                          e.stopPropagation();
                        }}
                        className="delete-button"
                      />
                    </Popconfirm>
                  </EmptyBox>
                </Space>
              </>
            )}
          </div>
          <EmptyBox if={!!errorMessage && deletingItem?.value === option.value}>
            <ReminderInformation status="error" message={errorMessage ?? ''} />
          </EmptyBox>
        </div>
      )
    }));

    return optionItems;
  };

  const dropdownContentRender = () => {
    // 自定义dropdownContent需要传递ref 否则antd会抛出ref warning
    const DropdownContentComponent = forwardRef<HTMLDivElement>(({}, ref) => (
      <Spin spinning={loading}>
        <EditableSelectStyleWrapper ref={ref as any} height={contentMaxHeight}>
          <EmptyBox if={searchable}>
            <div className="editable-select-search">
              <BasicInput
                value={searchValue}
                onChange={(e) => setSearchValue(e.target.value)}
                placeholder={
                  searchPlaceholder ??
                  t('common.actiontechTable.searchInput.placeholder')
                }
                allowClear
                bordered={false}
              />
            </div>
          </EmptyBox>

          <Menu
            className="editable-select-menu"
            items={generateMenuItems()}
            selectedKeys={value ? [value.toString()] : []}
          />

          <EmptyBox if={addable}>
            <div className="editable-select-add-section">
              {isAdding ? (
                <div className="edit-mode" onClick={(e) => e.stopPropagation()}>
                  <BasicInput
                    value={newItemName}
                    autoFocus
                    onChange={(e) => setNewItemName(e.target.value)}
                    placeholder={t('common.form.placeholder.input')}
                  />
                  <div className="button-group">
                    <BasicButton
                      size="small"
                      onClick={() => {
                        setIsAdding(false);
                        setNewItemName('');
                      }}
                    >
                      {t('common.cancel')}
                    </BasicButton>
                    <BasicButton
                      type="primary"
                      size="small"
                      onClick={handleAdd}
                      disabled={!newItemName.trim()}
                    >
                      {t('common.add')}
                    </BasicButton>
                  </div>
                </div>
              ) : (
                <Space
                  className="add-button-wraper"
                  onClick={() => startAdding()}
                >
                  <PlusOutlined width={14} height={14} />
                  {addButtonText ?? t('common.add')}
                </Space>
              )}
            </div>
          </EmptyBox>
        </EditableSelectStyleWrapper>
      </Spin>
    ));

    return <DropdownContentComponent />;
  };

  const selectedLabel = options.find((o) => o.value === value)?.label;

  return (
    <Dropdown
      dropdownRender={dropdownContentRender}
      trigger={['click']}
      open={dropdownOpen}
      onOpenChange={onOpenChange}
      disabled={disabled}
    >
      <EditableSelectTriggerStyleWrapper
        open={dropdownOpen}
        disabled={disabled}
        className={classNames('editable-select-trigger', {
          'editable-select-trigger-disabled': disabled
        })}
      >
        {selectedLabel ? (
          <span>{selectedLabel}</span>
        ) : (
          <span className="placeholder">
            {placeholder ?? t('common.form.placeholder.select')}
          </span>
        )}
        <span className="arrow-icon">
          {dropdownOpen ? (
            <UpOutlined width={20} height={20} />
          ) : (
            <DownOutlined width={20} height={20} />
          )}
        </span>
      </EditableSelectTriggerStyleWrapper>
    </Dropdown>
  );
};

export default EditableSelect;
