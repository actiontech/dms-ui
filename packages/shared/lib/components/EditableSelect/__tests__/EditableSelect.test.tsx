import { fireEvent, act, screen } from '@testing-library/react';
import { superRender } from '../../../testUtil/superRender';
import EditableSelect from '../EditableSelect';
import { EditableSelectOption } from '../index.type';
import {
  getBySelector,
  getAllBySelector,
  queryBySelector
} from '../../../testUtil/customQuery';

describe('EditableSelect', () => {
  const mockOptions: EditableSelectOption[] = [
    { value: '1', label: 'Option 1' },
    { value: '2', label: 'Option 2' }
  ];

  const mockOnChange = jest.fn();
  const mockOnAdd = jest.fn();
  const mockOnUpdate = jest.fn();
  const mockOnDelete = jest.fn();

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  const customRender = (props = {}) => {
    return superRender(
      <EditableSelect
        options={mockOptions}
        onChange={mockOnChange}
        onAdd={mockOnAdd}
        onUpdate={mockOnUpdate}
        onDelete={mockOnDelete}
        {...props}
      />
    );
  };

  it('should render select with options', async () => {
    const { baseElement } = customRender();
    expect(baseElement).toMatchSnapshot();

    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    expect(baseElement).toMatchSnapshot();
  });

  it('should render without options', async () => {
    const { baseElement } = customRender({ options: [] });
    expect(baseElement).toMatchSnapshot();
  });

  it('should render with loading status', async () => {
    customRender({ loading: true });
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    expect(getBySelector('.ant-spin')).toBeInTheDocument();
  });

  it('should render with selected value', async () => {
    customRender({ value: '1' });
    expect(getBySelector('.editable-select-trigger')?.textContent).toContain(
      'Option 1'
    );
  });

  it('should call onChange when select option', async () => {
    customRender();
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const firstOption = getAllBySelector('.ant-dropdown-menu-item')[0];
    fireEvent.click(firstOption);
    await act(async () => jest.advanceTimersByTime(0));

    expect(mockOnChange).toHaveBeenCalledWith('1');
  });

  it('should show add mode when click add button', async () => {
    customRender();
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const addButton = getAllBySelector('.ant-dropdown-menu-item')[2];
    fireEvent.click(addButton);
    await act(async () => jest.advanceTimersByTime(0));

    expect(getBySelector('.add-mode')).toBeInTheDocument();
  });

  it('should call onAdd when add new option', async () => {
    customRender();
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const addButton = getAllBySelector('.ant-dropdown-menu-item')[2];
    fireEvent.click(addButton);
    await act(async () => jest.advanceTimersByTime(0));

    const input = getBySelector('.add-mode input');
    fireEvent.change(input!, { target: { value: 'New Option' } });
    await act(async () => jest.advanceTimersByTime(0));

    const saveButton = getBySelector(
      '.add-mode .button-group button:last-child'
    );
    fireEvent.click(saveButton);
    await act(async () => jest.advanceTimersByTime(0));

    expect(mockOnAdd).toHaveBeenCalledWith('New Option');
  });

  it('should disable add button when input is empty', async () => {
    customRender();
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const addButton = getAllBySelector('.ant-dropdown-menu-item')[2];
    fireEvent.click(addButton);
    await act(async () => jest.advanceTimersByTime(0));

    expect(
      getBySelector('.add-mode .button-group button:last-child')
    ).toBeDisabled();
  });

  it('should show edit mode when click edit button', async () => {
    customRender();
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const editButton = getAllBySelector('.edit-button')[0];
    fireEvent.click(editButton);
    await act(async () => jest.advanceTimersByTime(0));

    expect(getBySelector('.edit-mode')).toBeInTheDocument();
  });

  it('should call onUpdate when edit option', async () => {
    customRender();
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const editButton = getAllBySelector('.edit-button')[0];
    fireEvent.click(editButton);
    await act(async () => jest.advanceTimersByTime(0));

    fireEvent.change(getBySelector('.edit-mode input'), {
      target: { value: 'Updated Option' }
    });
    await act(async () => jest.advanceTimersByTime(0));

    fireEvent.click(
      getBySelector('.edit-mode .button-group button:last-child')
    );
    await act(async () => jest.advanceTimersByTime(0));

    expect(mockOnUpdate).toHaveBeenCalledWith({
      value: '1',
      label: 'Updated Option'
    });
  });

  it('should disable edit button when input is empty', async () => {
    customRender();
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const editButton = getAllBySelector('.edit-button')[0];
    fireEvent.click(editButton);
    await act(async () => jest.advanceTimersByTime(0));

    fireEvent.change(getBySelector('.edit-mode input'), {
      target: { value: '' }
    });
    await act(async () => jest.advanceTimersByTime(0));
    expect(
      getBySelector('.edit-mode .button-group button:last-child')
    ).toBeDisabled();
  });

  it('should cancel edit when click cancel button', async () => {
    customRender();
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const editButton = getAllBySelector('.edit-button')[0];
    fireEvent.click(editButton);
    await act(async () => jest.advanceTimersByTime(0));

    fireEvent.click(
      getBySelector('.edit-mode .button-group button:first-child')
    );
    await act(async () => jest.advanceTimersByTime(0));

    expect(queryBySelector('.edit-mode')).not.toBeInTheDocument();
  });

  it('should call onDelete when confirm delete', async () => {
    customRender();
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const deleteButton = getAllBySelector('.delete-button')[0];
    fireEvent.click(deleteButton);
    await act(async () => jest.advanceTimersByTime(0));
    expect(screen.getByText('确认要删除么?')).toBeInTheDocument();

    const confirmButton = getBySelector('.ant-btn-primary.ant-btn-sm');
    fireEvent.click(confirmButton);
    await act(async () => jest.advanceTimersByTime(0));

    expect(mockOnDelete).toHaveBeenCalledWith(mockOptions[0]);
  });

  it('render error message when confirm delete failed', async () => {
    const errorMessage = '当前数据不可删除';
    customRender({
      errorMessage
    });
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const deleteButton = getAllBySelector('.delete-button')[0];
    fireEvent.click(deleteButton);
    await act(async () => jest.advanceTimersByTime(0));
    expect(screen.getByText('确认要删除么?')).toBeInTheDocument();

    const confirmButton = getBySelector('.ant-btn-primary.ant-btn-sm');
    fireEvent.click(confirmButton);
    await act(async () => jest.advanceTimersByTime(0));

    expect(mockOnDelete).toHaveBeenCalledWith(mockOptions[0]);

    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });

  it('should not show actions when disabled', async () => {
    customRender({
      disabled: true
    });

    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    expect(queryBySelector('.ant-dropdown')).not.toBeInTheDocument();
  });

  it('should not show add button when addable is false', async () => {
    customRender({
      addable: false
    });

    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    expect(getAllBySelector('.ant-dropdown-menu-item').length).toBe(
      mockOptions.length
    );
  });

  it('should not show edit button when updatable is false', async () => {
    customRender({
      updatable: false
    });

    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    expect(queryBySelector('.edit-button')).not.toBeInTheDocument();
  });

  it('should not show delete button when deletable is false', async () => {
    customRender({
      deletable: false
    });

    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    expect(queryBySelector('.delete-button')).not.toBeInTheDocument();
  });

  it('should reset state when dropdown closed', async () => {
    customRender();
    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const addButton = getAllBySelector('.ant-dropdown-menu-item')[2];
    fireEvent.click(addButton);
    await act(async () => jest.advanceTimersByTime(0));

    const firstOption = getAllBySelector('.ant-dropdown-menu-item')[0];
    fireEvent.click(firstOption);
    await act(async () => jest.advanceTimersByTime(0));

    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));
    expect(queryBySelector('.add-mode')).not.toBeInTheDocument();
  });

  it('should show custom placeholder', async () => {
    customRender({
      placeholder: 'Custom Placeholder'
    });

    expect(getBySelector('.placeholder')?.textContent).toBe(
      'Custom Placeholder'
    );
  });

  it('should show custom add button text', async () => {
    customRender({
      addButtonText: 'Custom Add'
    });

    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const addButton = getAllBySelector('.ant-dropdown-menu-item')[2];
    expect(addButton.textContent).toContain('Custom Add');
  });

  it('should show custom deletion confirm title', async () => {
    customRender({
      deletionConfirmTitle: 'Custom Delete Confirm'
    });

    fireEvent.click(getBySelector('.editable-select-trigger'));
    await act(async () => jest.advanceTimersByTime(0));

    const deleteButton = getAllBySelector('.delete-button')[0];
    fireEvent.click(deleteButton);
    await act(async () => jest.advanceTimersByTime(1000));
    expect(screen.getByText('Custom Delete Confirm')).toBeInTheDocument();
  });
});
