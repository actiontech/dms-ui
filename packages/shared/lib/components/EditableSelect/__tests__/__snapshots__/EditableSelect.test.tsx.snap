// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableSelect should render select when options is empty 1`] = `
<body>
  <div>
    <div
      class="ant-dropdown-trigger editable-select-trigger css-e92p61"
    >
      <span
        class="placeholder"
      >
        请选择{{name}}
      </span>
      <span
        class="arrow-icon"
      >
        <svg
          height="20"
          viewBox="0 0 16 16"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m8 8.781 3.3-3.3.943.943L8 10.667 3.757 6.424l.943-.943z"
            fill="#C3C6CD"
          />
        </svg>
      </span>
    </div>
  </div>
</body>
`;

exports[`EditableSelect should render select when options is empty 2`] = `
<body>
  <div>
    <div
      class="ant-dropdown-trigger editable-select-trigger ant-dropdown-open css-iiqyi4"
      open=""
    >
      <span
        class="placeholder"
      >
        请选择{{name}}
      </span>
      <span
        class="arrow-icon"
      >
        <svg
          fill="none"
          height="20"
          viewBox="0 0 16 16"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m8 7.219-3.3 3.3-.942-.943L8 5.333l4.243 4.243-.943.943z"
            fill="#C3C6CD"
          />
        </svg>
      </span>
    </div>
  </div>
  <div>
    <div
      class="ant-dropdown ant-slide-up-appear ant-slide-up-appear-prepare ant-slide-up ant-dropdown-placement-bottomLeft"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
    >
      <div
        class="css-8go0jz"
        height="320"
      >
        <div
          class="ant-spin-nested-loading"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="editable-select-search"
            >
              <span
                class="ant-input-affix-wrapper basic-input-wrapper css-1pd1cd5 ant-input-affix-wrapper-borderless"
              >
                <input
                  class="ant-input ant-input-borderless"
                  placeholder="输入关键字搜索"
                  type="text"
                  value=""
                />
                <span
                  class="ant-input-suffix"
                >
                  <span
                    class="ant-input-clear-icon ant-input-clear-icon-hidden"
                    role="button"
                    tabindex="-1"
                  >
                    <svg
                      fill="currentColor"
                      height="14"
                      viewBox="0 0 14 14"
                      width="14"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="m7 6.175 2.887-2.887.825.825L7.825 7l2.887 2.888-.825.824L7 7.825l-2.888 2.887-.824-.824L6.175 7 3.288 4.113l.824-.825z"
                      />
                    </svg>
                  </span>
                </span>
              </span>
            </div>
            <div
              class="ant-empty is-icon-tips css-19kysfv"
            >
              <div
                class="ant-empty-image"
              >
                <svg
                  fill="none"
                  height="80"
                  viewBox="0 0 80 80"
                  width="80"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    clip-path="url(#clip0_1412_15009)"
                  >
                    <path
                      d="M0 56.5643C0 61.4398 17.8769 65.3152 40.0043 65.3152C62.1316 65.3152 80.0086 61.4398 80.0086 56.5643C80.0086 51.6887 62.1317 47.8134 40.0043 47.8134C17.8768 47.8134 0 51.6888 0 56.5643Z"
                      fill="#EBEAE7"
                    />
                    <path
                      d="M52.0055 35.1869C52.0055 33.1867 53.2557 31.5615 54.7558 31.5615H68.7573V54.189C68.7573 56.8142 67.1322 59.0645 65.132 59.0645H15.0016C13.0014 59.0645 11.2512 56.9393 11.2512 54.189V31.5615H25.2527C26.7528 31.5615 28.003 33.1867 28.003 35.1869C28.003 37.1871 29.2531 38.8123 30.7533 38.8123H49.2553C50.7554 38.8123 52.0055 37.1871 52.0055 35.1869Z"
                      fill="#F2F1EF"
                    />
                    <path
                      d="M56.6311 16.56C56.0061 15.4349 55.0059 14.8098 53.6308 14.6848H26.3779C25.1277 14.8098 24.0026 15.4349 23.5026 16.56L10.6262 30.9365V54.189C10.6262 57.1893 12.5014 59.6896 15.0016 59.6896H65.007C67.3823 59.6896 69.2575 57.1893 69.2575 54.189V30.9365L56.6311 16.56ZM24.3777 17.3101C24.8777 16.435 25.5028 16.0599 26.3779 15.935H53.6308C54.5059 15.935 55.131 16.435 55.506 17.1851L67.6323 30.9365H54.7559C52.8807 30.9365 51.3805 32.8117 51.3805 35.187C51.3805 36.8122 50.3804 38.1873 49.2553 38.1873H30.7533C29.5032 38.1873 28.6281 36.8122 28.6281 35.187C28.6281 32.8117 27.128 30.9365 25.2528 30.9365H12.3763L24.3777 17.3101ZM68.1323 54.189C68.1323 56.4393 66.7572 58.4395 65.132 58.4395H15.0016C13.2515 58.4395 11.8763 56.5643 11.8763 54.189V32.1867H25.2528C26.3779 32.1867 27.378 33.5618 27.378 35.312C27.378 37.6873 28.8781 39.5625 30.7533 39.5625H49.2553C51.1306 39.5625 52.6307 37.6873 52.6307 35.312C52.6307 33.6868 53.6308 32.3117 54.7559 32.3117H68.0073V54.189H68.1323Z"
                      fill="#E6E4E3"
                    />
                  </g>
                  <defs>
                    <clippath
                      id="clip0_1412_15009"
                    >
                      <rect
                        fill="white"
                        height="80"
                        width="80"
                      />
                    </clippath>
                  </defs>
                </svg>
              </div>
              <div
                class="ant-empty-description"
              >
                <span
                  class="no-data-cont"
                >
                  暂无数据
                </span>
              </div>
              <div
                class="ant-empty-footer"
              />
            </div>
            <div
              class="editable-select-add-section"
            >
              <div
                class="ant-space ant-space-horizontal ant-space-align-center add-button-wraper"
              >
                <div
                  class="ant-space-item"
                  style="margin-right: 8px;"
                >
                  <svg
                    fill="none"
                    height="14"
                    viewBox="0 0 10 10"
                    width="14"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m5.02.5-.012 9M.5 5h9"
                      stroke="#1CB889"
                      stroke-linejoin="round"
                      stroke-width="1.5"
                    />
                  </svg>
                </div>
                <div
                  class="ant-space-item"
                >
                  添加
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`EditableSelect should render select with options 1`] = `
<body>
  <div>
    <div
      class="ant-dropdown-trigger editable-select-trigger css-e92p61"
    >
      <span
        class="placeholder"
      >
        请选择{{name}}
      </span>
      <span
        class="arrow-icon"
      >
        <svg
          height="20"
          viewBox="0 0 16 16"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m8 8.781 3.3-3.3.943.943L8 10.667 3.757 6.424l.943-.943z"
            fill="#C3C6CD"
          />
        </svg>
      </span>
    </div>
  </div>
</body>
`;

exports[`EditableSelect should render select with options 2`] = `
<body>
  <div>
    <div
      class="ant-dropdown-trigger editable-select-trigger ant-dropdown-open css-iiqyi4"
      open=""
    >
      <span
        class="placeholder"
      >
        请选择{{name}}
      </span>
      <span
        class="arrow-icon"
      >
        <svg
          fill="none"
          height="20"
          viewBox="0 0 16 16"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m8 7.219-3.3 3.3-.942-.943L8 5.333l4.243 4.243-.943.943z"
            fill="#C3C6CD"
          />
        </svg>
      </span>
    </div>
  </div>
  <div>
    <div
      class="ant-dropdown ant-slide-up-appear ant-slide-up-appear-prepare ant-slide-up ant-dropdown-placement-bottomLeft"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
    >
      <div
        class="css-8go0jz"
        height="320"
      >
        <div
          class="ant-spin-nested-loading"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="editable-select-search"
            >
              <span
                class="ant-input-affix-wrapper basic-input-wrapper css-1pd1cd5 ant-input-affix-wrapper-borderless"
              >
                <input
                  class="ant-input ant-input-borderless"
                  placeholder="输入关键字搜索"
                  type="text"
                  value=""
                />
                <span
                  class="ant-input-suffix"
                >
                  <span
                    class="ant-input-clear-icon ant-input-clear-icon-hidden"
                    role="button"
                    tabindex="-1"
                  >
                    <svg
                      fill="currentColor"
                      height="14"
                      viewBox="0 0 14 14"
                      width="14"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="m7 6.175 2.887-2.887.825.825L7.825 7l2.887 2.888-.825.824L7 7.825l-2.888 2.887-.824-.824L6.175 7 3.288 4.113l.824-.825z"
                      />
                    </svg>
                  </span>
                </span>
              </span>
            </div>
            <ul
              class="ant-dropdown-menu ant-dropdown-menu-root ant-dropdown-menu-vertical ant-dropdown-menu-light editable-select-menu"
              data-menu-list="true"
              role="menu"
              tabindex="0"
            >
              <li
                class="ant-dropdown-menu-item ant-dropdown-menu-item-only-child"
                data-menu-id="rc-menu-uuid-test-1"
                role="menuitem"
                tabindex="-1"
              >
                <span
                  class="ant-dropdown-menu-title-content"
                >
                  <div>
                    <div
                      class="menu-item-content"
                    >
                      <span>
                        Option 1
                      </span>
                      <div
                        class="ant-space ant-space-horizontal ant-space-align-center"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          <button
                            class="ant-btn ant-btn-text ant-btn-sm ant-btn-icon-only edit-button basic-button-wrapper css-geipcv"
                            type="button"
                          >
                            <span
                              class="ant-btn-icon"
                            >
                              <svg
                                height="16"
                                viewBox="0 0 16 16"
                                width="16"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M6.162 12.667H14V14H2v-2.829l6.6-6.6 2.828 2.83zm3.38-9.038 1.415-1.414a.667.667 0 0 1 .942 0L13.785 4.1a.667.667 0 0 1 0 .942l-1.414 1.414z"
                                  fill="#8A8F99"
                                />
                              </svg>
                            </span>
                          </button>
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <button
                            class="ant-btn ant-btn-text ant-btn-sm ant-btn-icon-only ant-btn-dangerous delete-button basic-button-wrapper css-geipcv"
                            type="button"
                          >
                            <span
                              class="ant-btn-icon"
                            >
                              <svg
                                height="14"
                                viewBox="0 0 18 18"
                                width="14"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M9 16.5a7.5 7.5 0 1 1 0-15 7.5 7.5 0 0 1 0 15M5.25 8.25v1.5h7.5v-1.5z"
                                  fill="#F66074"
                                />
                              </svg>
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </span>
              </li>
              <li
                class="ant-dropdown-menu-item ant-dropdown-menu-item-only-child"
                data-menu-id="rc-menu-uuid-test-2"
                role="menuitem"
                tabindex="-1"
              >
                <span
                  class="ant-dropdown-menu-title-content"
                >
                  <div>
                    <div
                      class="menu-item-content"
                    >
                      <span>
                        Option 2
                      </span>
                      <div
                        class="ant-space ant-space-horizontal ant-space-align-center"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          <button
                            class="ant-btn ant-btn-text ant-btn-sm ant-btn-icon-only edit-button basic-button-wrapper css-geipcv"
                            type="button"
                          >
                            <span
                              class="ant-btn-icon"
                            >
                              <svg
                                height="16"
                                viewBox="0 0 16 16"
                                width="16"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M6.162 12.667H14V14H2v-2.829l6.6-6.6 2.828 2.83zm3.38-9.038 1.415-1.414a.667.667 0 0 1 .942 0L13.785 4.1a.667.667 0 0 1 0 .942l-1.414 1.414z"
                                  fill="#8A8F99"
                                />
                              </svg>
                            </span>
                          </button>
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <button
                            class="ant-btn ant-btn-text ant-btn-sm ant-btn-icon-only ant-btn-dangerous delete-button basic-button-wrapper css-geipcv"
                            type="button"
                          >
                            <span
                              class="ant-btn-icon"
                            >
                              <svg
                                height="14"
                                viewBox="0 0 18 18"
                                width="14"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M9 16.5a7.5 7.5 0 1 1 0-15 7.5 7.5 0 0 1 0 15M5.25 8.25v1.5h7.5v-1.5z"
                                  fill="#F66074"
                                />
                              </svg>
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </span>
              </li>
            </ul>
            <div
              aria-hidden="true"
              style="display: none;"
            />
            <div
              class="editable-select-add-section"
            >
              <div
                class="ant-space ant-space-horizontal ant-space-align-center add-button-wraper"
              >
                <div
                  class="ant-space-item"
                  style="margin-right: 8px;"
                >
                  <svg
                    fill="none"
                    height="14"
                    viewBox="0 0 10 10"
                    width="14"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m5.02.5-.012 9M.5 5h9"
                      stroke="#1CB889"
                      stroke-linejoin="round"
                      stroke-width="1.5"
                    />
                  </svg>
                </div>
                <div
                  class="ant-space-item"
                >
                  添加
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`EditableSelect should render without options 1`] = `
<body>
  <div>
    <div
      class="ant-dropdown-trigger editable-select-trigger css-e92p61"
    >
      <span
        class="placeholder"
      >
        请选择{{name}}
      </span>
      <span
        class="arrow-icon"
      >
        <svg
          height="20"
          viewBox="0 0 16 16"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m8 8.781 3.3-3.3.943.943L8 10.667 3.757 6.424l.943-.943z"
            fill="#C3C6CD"
          />
        </svg>
      </span>
    </div>
  </div>
</body>
`;
